ddl/partition.go

```diff
@@ -1583,20 +1583,32 @@ func onAlterTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
 		return 0, errors.Trace(table.ErrUnknownPartition.GenWithStackByArgs("drop?", tblInfo.Name.O))
 	}
 
-	err = infosync.PutRuleBundles(nil, []*placement.Bundle{bundle})
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return 0, errors.Wrapf(err, "failed to notify PD the placement rules")
-	}
-
-	// used by ApplyDiff in updateSchemaVersion
-	job.CtxVars = []interface{}{partitionID}
-	ver, err := updateSchemaVersion(t, job)
-	if err != nil {
-		return ver, errors.Trace(err)
+	pstate := ptInfo.GetStateByID(partitionID)
+	switch pstate {
+	case model.StatePublic:
+		err = infosync.PutRuleBundles(nil, []*placement.Bundle{bundle})
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return 0, errors.Wrapf(err, "failed to notify PD the placement rules")
+		}
+		ptInfo.SetStateByID(partitionID, model.StateGlobalTxnOnly)
+		// used by ApplyDiff in updateSchemaVersion
+		job.CtxVars = []interface{}{partitionID}
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StateGlobalTxnOnly
+	case model.StateGlobalTxnOnly:
+		ptInfo.SetStateByID(partitionID, model.StatePublic)
+		// used by ApplyDiff in updateSchemaVersion
+		job.CtxVars = []interface{}{partitionID}
```

Comment:
I think you can put the bundles handling logic & job parameter change into the StateGlobalTxnOnly.
Because:
there is a possibility that: PD receives the changing roles after HTTP request in line 1589, then began scheduling the partition, while some TiDB instance hasn't notified with `StateGlobalTxnOnly` state change, global txn can still happen.

Reply:
+1 for notifying PD in `StateGlobalTxnOnly`. It makes sense to me.

Reply:
nice catch.

Reply:
If PD start schedule in GlobalTxn, PD will complete schedule before `StateGlobalTxn->StatePublic`, is that true?

Reply:
> If PD start schedule in StateGlobalTxn, PD will complete schedule before StateGlobalTxn->StatePublic, is that true?

Possibly.

comment addressed.